core-api/src/telemetry/metrics.rs

```diff
@@ -0,0 +1,263 @@
+    }
+
+    fn gauge(&self, name: &str) -> Arc<dyn Gauge> {
+        self.as_ref().gauge(name)
+    }
+}
+
+/// Attributes which are provided every time a call to record a specific metric is made.
+/// Implementors must be very cheap to clone, as these attributes will be re-used frequently.
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub enum MetricAttributes {
+    #[cfg(feature = "otel_impls")]
+    OTel {
+        kvs: Arc<Vec<opentelemetry::KeyValue>>,
+        ctx: opentelemetry::Context,
+    },
+    Lang {
+        /// An opaque reference to attributes stored in lang memory
+        id: u64,
+        /// If populated, these key values should also be used in addition to the referred-to
+        /// existing attributes when recording
+        new_attributes: Vec<MetricKeyValue>,
+    },
+}
+
+impl MetricAttributes {
+    /// Extend existing metrics attributes with new k/vs, returning a new instance
+    pub fn with_new_attrs(&self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) -> Self {
+        let mut me = self.clone();
+        me.add_new_attrs(new_kvs);
+        me
+    }
+
+    /// Mutate self to add new kvs
+    pub fn add_new_attrs(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        match self {
+            #[cfg(feature = "otel_impls")]
+            MetricAttributes::OTel { ref mut kvs, .. } => {
+                Arc::make_mut(kvs).extend(new_kvs.into_iter().map(Into::into));
+            }
+            MetricAttributes::Lang {
+                ref mut new_attributes,
+                ..
+            } => {
+                new_attributes.extend(new_kvs.into_iter());
+            }
+        }
+    }
+}
+
+/// Options that are attached to metrics on a per-call basis
+#[derive(Clone, Default, derive_more::Constructor)]
+pub struct MetricsAttributesOptions {
+    attributes: Vec<MetricKeyValue>,
+}
+impl MetricsAttributesOptions {
+    pub fn extend(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        self.attributes.extend(new_kvs.into_iter())
+    }
+}
+
+/// A K/V pair that can be used to label a specific recording of a metric
+#[derive(Clone, Debug)]
+pub struct MetricKeyValue {
+    key: String,
+    value: MetricValue,
+}
+impl MetricKeyValue {
+    pub fn new(key: impl Into<String>, value: impl Into<MetricValue>) -> Self {
+        Self {
+            key: key.into(),
+            value: value.into(),
+        }
+    }
+}
+
+/// Values metric labels may assume
+#[derive(Clone, Debug, derive_more::From)]
+pub enum MetricValue {
+    String(String),
+    Int(i64),
+    Float(f64),
+    Bool(bool),
+    // can add array if needed
+}
+impl From<&'static str> for MetricValue {
+    fn from(value: &'static str) -> Self {
+        MetricValue::String(value.to_string())
+    }
+}
+
+/// Histogram buckets
+#[derive(Clone, Debug)]
+pub struct Buckets {
+    buckets: Vec<f64>,
+}
+
+pub trait Counter: Send + Sync {
+    fn add(&self, value: u64, attributes: &MetricAttributes);
```

Comment:
Providing attributes at metric record time can get a bit expensive for naive implementations that need to group by collective attributes. There is a cost associated with not pre-binding attributes. Can we consider an alteration in the API (even if the primary impl of that doesn't use it that way) that associates metric attributes with a `CoreMeter` instead? And you can have `with_new_attrs` that returns `Self` on that instead? Or at least have `with_new_attrs` on each of these metrics instead? This way implementers can implement these metric impls having already created the counter with its attributes. And it also discourages users of these metrics from adding any attribute willy-nilly at record time.

Reply:
Having thought about this, I now fear that the way the metrics are recorded may preclude this. While `counter.with_new_attrs(attrs).add(123)` is just as cheap as `counter.add(123, attrs)`, if you are always calling `with_new_attrs` on every record time anyways then there's no value here. So this can be ignored.

Just know that many non-OTel implementations are required to group by these attributes before applying the value. But otherwise, this can be ignored.

Reply:
Looks like this will work with the Otel JS metrics API: https://github.com/open-telemetry/opentelemetry-js/blob/b8b63083869fb5850fd4830ded0fc509a2cd0f0a/api/src/metrics/Metric.ts#L65

Reply:
We'll lose the otel context but other than that everything else should work the way it is.

Reply:
That lang controls creating this set (e.g. w/ pre-built hashing) does seem to help, so long as the Rust side doesn't commonly create this attribute set on each recording (here and there can make sense though).

[Go Prometheus](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus) for instance takes the labels (what we call attributes here) and uses fnv1-a to hash them for grouping purposes, but is optimized to do that ahead of time if the caller does it ahead of time. And so they don't take labels at record time to discourage dynamic labelling.

Reply:
I'm missing how lang is supposed to maintain this cache, aren't most attributes defined core side?